consensus/bor/api.go

```diff
@@ -105,3 +120,75 @@ func (api *API) GetCurrentValidators() ([]*Validator, error) {
 	}
 	return snap.ValidatorSet.Validators, nil
 }
+
+// GetRootHash returns the merkle root of the start to end block headers
+func (api *API) GetRootHash(start int64, end int64) ([]byte, error) {
+	key := getKey(start, end)
+	if root, err := api.lookupCache(key); err != nil {
+		return nil, err
+	} else if root != nil {
+		return root, nil
+	}
+	length := uint64(end - start + 1)
+	if length > MaxCheckpointLength {
+		return nil, &MaxCheckpointLengthExceededError{start, end}
+	}
+	currentHeaderNumber := api.chain.CurrentHeader().Number.Int64()
+	if start > end || end > currentHeaderNumber {
+		return nil, &InvalidStartEndBlockError{start, end, currentHeaderNumber}
+	}
+	blockHeaders := make([]*types.Header, end-start+1)
+	wg := new(sync.WaitGroup)
+	concurrent := make(chan bool, 20)
+	for i := start; i <= end; i++ {
+		wg.Add(1)
+		concurrent <- true
+		go func(number int64) {
+			blockHeaders[number-start] = api.chain.GetHeaderByNumber(uint64(number))
+			<-concurrent
+			wg.Done()
+		}(i)
+	}
+	wg.Wait()
+	close(concurrent)
+
+	headers := make([][32]byte, nextPowerOfTwo(length))
+	for i := 0; i < len(blockHeaders); i++ {
+		blockHeader := blockHeaders[i]
+		header := crypto.Keccak256(appendBytes32(
+			blockHeader.Number.Bytes(),
+			new(big.Int).SetUint64(blockHeader.Time).Bytes(),
+			blockHeader.TxHash.Bytes(),
+			blockHeader.ReceiptHash.Bytes(),
+		))
+
+		var arr [32]byte
+		copy(arr[:], header)
+		headers[i] = arr
+	}
+
+	tree := merkle.NewTreeWithOpts(merkle.TreeOptions{EnableHashSorting: false, DisableHashLeaves: true})
+	if err := tree.Generate(convert(headers), sha3.NewLegacyKeccak256()); err != nil {
+		return nil, err
+	}
+	root := tree.Root().Hash
+	api.rootHashCache.Add(key, root)
```

Comment:
`api.rootHashCache` can be nil. Let's create `getRootHashCache` which uniquely creates a root hash instance. Something like this: https://github.com/maticnetwork/heimdall/blob/c5aeaa63508219af573b64035265e4f8aace00a9/auth/types/pulp.go#L24-L33

Reply:
Done.